### PR TITLE
Swift: Fix for MaD with generics

### PR DESCRIPTION
**Pull request by geoffw0**
Fix for models-as-data matching CSV rows with generic class member functions or fields, a problem I ran into on the Alamofire taint source work.  In fact there are (at least) three interacting features we have to consider - generics, subclasses (including protocols) and extensions.  I've added a test that attempts to cover combinations of all three.

Working with `TypeDecl`s rather than `Type`s for as long as possible in `interpretElement0` seems to resolve the issue with generics at no cost to the other two features.

### CODE REVIEW

File: swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll [hunk 1]
```diff
@@ -390,16 +390,12 @@ predicate matchesSignature(AbstractFunctionDecl func, string signature) {
   paramsString(func) = signature
 }
 
-private NominalType getDeclType(IterableDeclContext decl) {
-  result = decl.(ClassDecl).getType()
+private NominalTypeDecl resolveExtensions(IterableDeclContext decl) {
+  // TODO: this should be a method on IterableDeclContext
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll:394]: If we're happy with the approach I will move this predicate into `IterableDeclContext`.

File: swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll [hunk 2]
```diff
@@ -390,16 +390,12 @@
...
+  result = decl.(NominalTypeDecl)
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll:395]: Dismissed because `.(NominalTypeDecl)` makes it clearer what this case is doing.

**Review by MathiasVP — COMMENTED** [review]
LGTM! We should probably do a DCA run before we merge this, though.

**Comment by geoffw0** [comment]
Yep, and I'll fix the TODOs (by doing the first and turning the second into an issue).

**Comment by geoffw0** [comment]
Issue for extensions with protocols: https://github.com/github/codeql-c-team/issues/1435

File: swift/ql/lib/codeql/swift/elements/decl/IterableDeclContext.qll [hunk 3]
```diff
@@ -1,4 +1,18 @@
...
-class IterableDeclContext extends Generated::IterableDeclContext { }
+/**
+ * A nominal type (class, struct, enum or protocol) or extension.
+ */
+class IterableDeclContext extends Generated::IterableDeclContext {
+  /**
+   * Gets the `NominalTypeDecl` corresponding to this `IterableDeclContext`
+   * resolving an extension to the extended type declaration.
+   */
+  NominalTypeDecl resolveExtensions() {
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/elements/decl/IterableDeclContext.qll:13]: I'm not sure I like the current name. Could we maybe use a name that's more consistent with the QL style-guide (i.e., `getX`)? Maybe something like `getResolvedTypeDecl`?

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/elements/decl/IterableDeclContext.qll:13]: Yeah, I see what you mean...

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/elements/decl/IterableDeclContext.qll]: We also use `resolved` a lot to mean something different, so its looking like `resolveExtensions` was overall a pretty poor name.  I've gone with `getNominalTypeDecl` for now, does that sound OK to you?

Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/elements/decl/IterableDeclContext.qll]: That sounds okay to me, yeah 👍

File: swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll [hunk 4]
```diff
@@ -390,16 +390,12 @@ predicate matchesSignature(AbstractFunctionDecl func, string signature) {
...
   or
-  result = decl.(StructDecl).getType()
-  or
-  result = getDeclType(decl.(ExtensionDecl).getExtendedTypeDecl())
-  or
-  result = decl.(EnumDecl).getType()
-  or
-  result = decl.(ProtocolDecl).getType()
+  result = decl.(ExtensionDecl).getExtendedTypeDecl()
+  // TODO: or a protocol added by the extension
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll]: This is now an issue here: https://github.com/github/codeql-c-team/issues/1435

File: swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll (hunk 1 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll]: Done.

**Comment by geoffw0** [comment]
DCA LGTM.

**Review by MathiasVP — APPROVED** [review]

**Pull request merged**